pytorch_lightning/trainer/connectors/accelerator_connector.py

```diff
@@ -57,975 +56,801 @@
+            elif len(self._parallel_devices) <= 1:
+                device = (
+                    device_parser.determine_root_gpu_device(self._parallel_devices)
+                    if self._accelerator_flag == "gpu"
+                    else "cpu"
+                )
+                # TODO: lazy initialized device, then here could be self._strategy_flag = "single_device"
+                self._strategy_flag = SingleDeviceStrategy(device=device)
+            elif len(self._parallel_devices) > 1:
+                self._strategy_flag = "ddp_spawn"
+            else:
+                self._strategy_flag = "ddp"
 
-    @property
-    def root_gpu(self) -> Optional[int]:
-        return (
-            self.strategy.root_device.index
-            if not isinstance(self.accelerator, (IPUAccelerator, TPUAccelerator))
-            else None
-        )
+    def _check_strategy_and_fallback(self) -> None:
+        """Checks edge cases when the strategy selection was a string input, and we need to fall back to a
+        different choice depending on other parameters or the environment."""
+        # current logic, fallback only apply to user pass in str config not object config
+        strategy_flag = "" if isinstance(self._strategy_flag, Strategy) else self._strategy_flag
 
-    @staticmethod
-    def _is_plugin_training_type(plugin: Union[str, Strategy]) -> bool:
-        if isinstance(plugin, str) and (plugin in StrategyRegistry or plugin in list(_StrategyType)):
-            return True
-        return isinstance(plugin, Strategy)
+        if strategy_flag == "ddp_cpu":
+            if _TPU_AVAILABLE:
+                raise MisconfigurationException(
+                    "`accelerator='ddp_cpu'` is not supported on TPU machines. "
+                    "Learn more: https://github.com/PyTorchLightning/pytorch-lightning/issues/7810"
+                )
+            if self._device_flag == 1 and self._num_nodes_flag > 1:
+                strategy_flag = "ddp"
+            else:
+                strategy_flag = "ddp_spawn"
+            if self._accelerator_flag == "gpu":
+                rank_zero_warn(
+                    "You requested one or more GPUs, but set `accelerator='ddp_cpu'`. Training will not use GPUs."
+                )
+        if strategy_flag in ("ddp_spawn", "ddp_spawn_find_unused_parameters_false") and (
+            TorchElasticEnvironment.detect() or KubeflowEnvironment.detect() or self._is_slurm_managing_tasks()
+        ):
+            strategy_flag = "ddp"
+        if strategy_flag in ("dp", "ddp2") and self._accelerator_flag == "cpu":
+            rank_zero_warn(f"{strategy_flag!r} is not supported on CPUs, hence setting `strategy='ddp'`.")
+            strategy_flag = "ddp"
 
-    @property
-    def is_training_type_in_plugins(self) -> bool:
-        return any(
-            (isinstance(plug, str) and plug in StrategyRegistry) or isinstance(plug, Strategy) for plug in self.plugins
-        )
+        if strategy_flag:
```

Comment:
I would override directly self._strategy_flag instead of making it lazy like this.

Reply:
It's more readable and easier to debug this way I think, otherwise we need to check/assign every step.